helper/schema/resource_diff.go

```diff
@@ -0,0 +1,453 @@
+package schema
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/terraform/terraform"
+)
+
+// newValueWriter is a minor re-implementation of MapFieldWriter to include
+// keys that should be marked as computed, to represent the new part of a
+// pseudo-diff.
+type newValueWriter struct {
+	*MapFieldWriter
+
+	// A list of keys that should be marked as computed.
+	computedKeys map[string]bool
+
+	// A lock to prevent races on writes. The underlying writer will have one as
+	// well - this is for computed keys.
+	lock sync.Mutex
+}
+
+// WriteField overrides MapValueWriter's WriteField, adding the ability to flag
+// the address as computed.
+func (w *newValueWriter) WriteField(address []string, value interface{}, computed bool) error {
+	if err := w.MapFieldWriter.WriteField(address, value); err != nil {
+		return err
+	}
+
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	if w.computedKeys == nil {
+		w.computedKeys = make(map[string]bool)
+	}
+
+	if computed {
+		w.computedKeys[strings.Join(address, ".")] = true
```

Comment:
Hmm, I was running under the assumption that a zero value could still be legitimately zero and not computed, probably a mistake I made by looking at what `NewComputed` diffs look like. Will fix for sure - we will also need to adjust `SetNewComputed` so that it doesn't try and pull a zero value for the key in the schema, and just sets a nil value. Does that sound right?